MT_toolbox/spectrum_lib.py

```diff
@@ -0,0 +1,262 @@
+    else:
+        if output_level > 0:
+            print('\t[INFO] Convergence found.')
+        return [m[0][0], m[1][0]]
+
+
+def new_weights(abs_error, scale):
+    """
+
+    # Return weights depending on error scale
+    # Input:
+    # abs_error = dot product of spread in electric field between measured and predicted
+    # scale = here, normalized MAD of the error on electric field
+    # Output:
+    # Weights to be applied to the system
+    """
+    weights = []
+    for sample_error in abs_error:
+        if sample_error[0] / scale <= 1.5:
+            weights.append(1)
+        else:
+            weights.append(1.5 / np.abs(sample_error[0] / scale))
+    #weights = np.reshape(weights, (len(weights), 1))
+
+    return weights
+
+def error_on_output(m, d, G, weights):
+    """
+
+    # Return error and scale information for the robust regression
+    # Input:
+    # m: model (ie Z) in the Gm = d problem to solve
+    # d: data (ie electric field)
+    # G: Predictor (ie magnetic field, 2xN matrix)
+    # weights: applied weights to the system
+    # Output:
+    # scale: MAD/0.44845 to normalize it (ie make it consistent with a std)
+    # abs_error: d - Gm
+    # sum_error: dot product of abs_error with weights.
+    """
+
+    error = np.asarray(d - np.matrix(G) * np.matrix(m))
+
+    abs_error = np.abs(np.conjugate(error) * error)
+    scale = np.median(np.abs(abs_error - np.median(abs_error))) / 0.44845
+    if scale == 0:
+        return np.asarray([[np.nan, np.nan]])
+
+    weights = np.reshape(np.asarray(weights), (len(weights), 1))
+    abs_error_weighted = np.abs(np.conjugate(error) * np.asarray(weights) * np.asarray(error))
+    sum_error = np.dot(abs_error_weighted[:, 0], abs_error_weighted[:, 0])
+
+    return scale, abs_error, sum_error
+
+
+def organize_matrices(output, input, ref, weights):
+    """
+
+    # Return matrices for linear inversion
+    # Input:
+    # output: output channels in transfer function system (ie. for MT, electric field)
+    # input: input channels in transfer function system (ie. for MT, magnetic field)
+    # ref: reference channels in transfer function system (ie. for MT, reference magnetic field)
+    # weights: applied weights to the system
+    # Output:
+    # G: 2xN matrix of the local magnetic field
+    # GR: 2xN matrix of the remote magnetic field
+    # d: 1xN matrix of the electric field
+    # weights_matrix: (NxN) diagonal matrix with weights.
+    """
+
+    n_segments = len(output)
+    G = np.asarray([[input[i][j] for i in range(len(input))] for j in range(n_segments)])
+    GR = np.asarray([[ref[i][j] for i in range(len(ref))] for j in range(n_segments)])
+    d = np.asarray([[output[j]] for j in range(n_segments)])
+
+
+    weights_matrix = [[0. for j in range(n_segments)] for i in range(n_segments)]
+    for i in range(n_segments):
+        weights_matrix[i][i] = weights[i]
+
+    return G, GR, d, weights_matrix
+
+def invert_tensor(G, GR, d, weights_matrix):
```

Comment:
I assume this time consuming function?

Reply:
The invert_tensor is the time consuming one yes. 
Line 260: m = np.asarray(slin.solve(cm, d))

I had the same problem with my wavelet code, so I wrote it in Fortran using the LAPACK library, but since you're the C man, I think we could just rewrite it in C. I'm not a big connoisseur of C library for linear algebra.
Either rewrite only the inversion part, or directly the whole robust regression part (there's one more part to it coming, using the predictor hat matrix)